🚀 4단계 - 수강신청(요구사항 변경)

Perhona opened the pull request:
석홍님 안녕하세요~ 4단계 구현해보았습니다.

이번에 구현한 시나리오는 다음과 같습니다.

1. NsUserSessions는 강의 & 선발된유저 목록을 가진다.
2. Teacher는 수강신청 내용을 확인하고 선발 유저 목록에 기반하여 승인 혹은 거절한다.

선발된 인원이 있어 현재 수강 인원과 최대 수강 인원을 관리할 필요가 없다고 생각하여, 관련 내용을 제거하였습니다.

참고해서 봐주시면 감사하겠습니다! 😄

Perhona committed:
feat(step4): 강의는 하나 이상의 커버 이미지를 가진다.

Perhona committed:
feat(step4): 강의가 모집중인 상태에도 수강신청이 가능하도록 함

Perhona committed:
feat(step4): 강사는 선발된 인원에 대해서만 수강 승인

- 선발 인원이 정해져 있으므로 userNumber관련 메서드 삭제
- Enrollment 삭제

seokhong commented on src/main/java/nextstep/courses/domain/session/SessionCondition.java:
```diff
@@ -17,13 +17,10 @@ public SessionCondition(Long amount, Long maxUserNumber) {
         if (!payment.isSameAmount(amount)) {
             throw new CannotEnrollException("결제 금액이 일치하지 않습니다.");
         }
-        if (isPaidSession()) {
```
수강 승인을 할 수 있는 인원이 최대 인원으로 제한되도록 변경해보는 건 어떨까요?
또한, 이미 수강 제한 인원만큼 승인이 되었다면, 추가로 수강 신청을 할 수 없도록 제한할 필요도 있지 않을까요?

seokhong commented on src/main/java/nextstep/courses/domain/session/SessionStatus.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    public void canEnroll() throws CannotEnrollException {
+        if (!sessionRecruitmentStatus.isRecruiting()) {
```
4단계의 프로그래밍 요구사항을 보시면 아래와 같은 내용이 있습니다.
* DB 테이블에 데이터가 존재한다는 가정하에 리팩터링해야 한다.
  * 즉, 기존에 쌓인 데이터를 제거하지 않은 상태로 리팩터링 해야 한다.

기능 요구사항이 변경되기 전 기존의 데이터에는 `sessionProgressStatus`만 있을텐데요. `sessionProgressStatus`만 있는 경우에도 수강 신청이 가능하도록(기존과 동일하게 동작하도록) 구현해보시면 좋겠습니다. 이러한 데이터는 `sessionRecruitmentStatus`가 없기 때문에, 현재 코드에서는 NPE가 발생할 것으로 보입니다.

seokhong commented on src/main/java/nextstep/courses/domain/session/SessionProgressStatus.java:
```diff
@@ -0,0 +1,17 @@
+package nextstep.courses.domain.session;
+
+public enum SessionProgressStatus {
```
기존의 상태를 보면 아래와 같은 값을 가질 수 있네요. 프로그래밍 요구사항에 따라 기존 데이터가 존재하므로, `RECRUITING` 상태의 데이터가 남아 있을 것으로 보입니다. 기존 데이터에 대해서도 잘 동작하도록 구현해보시면 좋겠습니다.
* PREPARING, RECRUITING, CLOSED

seokhong commented on src/main/java/nextstep/courses/service/SessionService.java:
```diff
@@ -18,18 +21,16 @@ public class SessionService {
 
     public void enroll(Payment payment) throws CannotEnrollException {
-        List<NsUserSession> nsUserSessions = sessionDAO.getNsUserSessions(payment.getSessionId());
+        NsUserSessions nsUserSessions =new NsUserSessions(sessionDAO.findNsUserSessionsBySessionId(payment.getSessionId()));
```
`=` 과 `new` 키워드 사이에 공백이 누락됐네요.
IDE의 자동 줄맞춤 단축키를 사용하시거나, 파일을 저장할 때마다 자동으로 줄맞춤이 되도록 설정해보시면 좋겠습니다.
intellij를 사용하신다면 Preferences > Tools > Actions on Save 에서 자바 코드도 Reformat Code, Optimize Imports를 설정해서 수정하는 코드를 자동 줄맞춤, 임포트 최적화를 실행하도록 하실 수 있습니다. :)

seokhong replied:
`enroll` 메서드는 수강 신청을 하는 기능이므로, `Teacher`가 수강 신청을 승인/취소하는 것은 별도의 기능으로 분리되어야 할 것으로 보입니다.  `enroll` 메서드는 기존과 동일하게 `NsUser`가 `Session`에 대한 수강 신청(`NsUserSession` 생성)을 하면 될 것으로 보입니다.

seokhong commented on src/main/java/nextstep/courses/domain/NsUserSession.java:
```diff
@@ -3,25 +3,33 @@
-    private Long nsUserId;
+    private long sessionId;
+    private long nsUserId;
+    private boolean registered;
```
`NsUserSession`은 3가지 상태를 가질 수 있을 것으로 보입니다.
* 신청, 승인, 취소

상태를 관리할 수 있도록 enum 으로 변경해보면 어떨까요?

seokhong commented on src/main/java/nextstep/courses/domain/session/Session.java:
```diff
@@ -68,11 +56,12 @@ private void validate(Long courseId) {
+        sessionStatus.canEnroll();
+        sessionCondition.match(payment);
+
+        return new NsUserSession(payment.getSessionId(), payment.getNsUserId(), true);
```
강의가 생성된 시점에 따라, 승인 기능이 생성되기 전 강의라면 수강 신청을 하는 순간 승인이 될 것으로 보이고, 승인 기능이 만들어진 후에 생성된 강의라면 수강 신청의 상태가 되지 않을까요?

seokhong commented on src/main/resources/schema.sql:
기존 데이터가 존재하는 상황에서 기능 요구사항이 변경된 것이므로, create table DDL 문을 수정하지 마시고, alter table 을 이용해서 테이블 구조를 변경하는 형태로 진행해보시면 좋겠습니다.
`session` 테이블의 `session_status`도 유지되어야 하므로 삭제하면 안 될 것으로 보이네요!

seokhong commented on src/main/java/nextstep/users/domain/Teacher.java:
```diff
@@ -0,0 +1,17 @@
+
+import nextstep.courses.domain.NsUserSessions;
+
+public class Teacher {
```
강사가 수강 신청을 승인/취소 하는 기능이 추가되기 위해서는 `Session`에 강사 정보가 추가되고 검증도 해야 할 것으로 보입니다.
`NsUser`가 사용자 정보를 가지고 있으므로, `Session`의 강사도 `NsUser`를 이용해 저장할 수 있을 것으로 보입니다.

seokhong reviewed (changes requested):
남은님, 4단계 미션 잘 구현하셨습니다. 👍 
코드에 코멘트 남겼습니다. 코멘트 확인 부탁드립니다. 🙂

Perhona committed:
docs(step4): README.md 피드백 사항 추가

Perhona committed:
refactor(step4): 기존의 강의를 삭제하지 않으면서 강의 상태를 분류한다.

- 변경했던 session 테이블의 create 구문을 되돌리고, alter 구문 추가
- SessionProgressStatus에 enum 값 empty를 추가하고, 해당 컬럼 값이 존재하지 않는 기존 강의의 값을 제어하도록 함

Perhona committed:
refactor(step4): 수강 신청 정보가 신청 상태값을 갖도록 변경

- nsUserSession enrollment_status 컬럼 추가
- 기능 구현 전의 데이터를 관리하기 위해, session approval_required 컬럼 추가
- 추후 수강인원 확인을 위해 session이 nsUserSessions를 알고있도록 변경

Perhona committed:
refactor(step4): 수강 승인 기능을 별도로 분리

- Teacher 클래스 삭제
- 수강신청 승인, 취소 테스트 추가

Perhona commented:
포비님 안녕하세요~!

리뷰 사항이 다소 어렵게 느껴져 늦었지만 수정하여 리뷰 요청드립니다.
이전 단계들과는 달리 실무적인 코드이다 보니, 지금껏 배웠던 인터페이스 활용, 메서드, 클래스 분리 등을 적용하는 것에 어려움이 있었던 것 같습니다.

제 코드의 시나리오는 결제 정보로 수강 신청 (NsUserSession)을 진행하고,
강사가 신청 정보를 변경하면 강사 정보 및 수강 신청 정보로 승인 혹은 취소를 진행하는 흐름입니다.

추가로, LMS 코드에서 인터페이스나 디자인 패턴을 활용할 수 있는 곳은 어디인지 궁금합니다.

새해 복 많이 받으세요 ! 😄 
감사합니다.

javajigi commented on README.md:
```diff
@@ -101,3 +101,40 @@
+  * ns_user_session 테이블에 enrollment_status 컬럼 추가, default value 'APPROVED'
+* [x] Session에 강사 정보를 추가 및 검증한다.
+* [x] 수강 승인을 할 수 있는 인원이 최대 수강 인원으로 제한되도록 해본다.
+  * 추가로, 수강 인원만큼 승인이 되었다면 추가로 수강 승인이 되지 않도록 제한한다.
```
todo list 정리 👍

javajigi commented on src/main/java/nextstep/courses/domain/NsUserSession.java:
```diff
@@ -3,32 +3,60 @@
 import java.util.Objects;
 
 public class NsUserSession {
```
이 이름도 괜찮지만 이는 NsUser와 Session을 매핑하는 개발자 관점에서의 이름이지 않을까?
도메인 입장에서의 더 적합한 이름은 Student와 같은 이름이지 않을까?

javajigi commented on src/main/java/nextstep/courses/domain/NsUserSession.java:
```diff
@@ -3,32 +3,60 @@
+    }
+
+    public boolean isApproved() {
+        return enrollmentStatus == EnrollmentStatus.APPROVED;
```
enum도 객체라 `enrollmentStatus.isApproved()`와 같이 enum에 위임하는 것도 한 가지 방법임
단, 좀 과하다 생각할 수도 있음.

javajigi commented on src/main/java/nextstep/courses/domain/NsUserSession.java:
```diff
@@ -3,32 +3,60 @@
+    public boolean isApproved() {
+        return enrollmentStatus == EnrollmentStatus.APPROVED;
     }
 
```
수강 승인(approve), 취소(cancel)을 담당하는 메서드도 이 객체가 가지는 것은 어떨까?
즉 자신의 수강 상태를 변경하는 로직을 자신이 가진다.

javajigi commented on src/main/java/nextstep/courses/domain/NsUserSessions.java:
```diff
@@ -1,15 +1,27 @@
+import java.util.List;
+
+public class NsUserSessions {
+    private List<NsUserSession> nsUserSessions;
```
대부분의 필드에 final을 추가하지 않고 있다.
값을 변경할 수 없는 필드에 대해 명시적으로 final 키워드를 추가하는 습관을 들이는 것은 어떨까?

javajigi commented on src/main/java/nextstep/courses/domain/NsUserSessions.java:
```diff
@@ -1,15 +1,27 @@
+    public Long approvedUserNumber() {
+        return nsUserSessions.stream().filter(NsUserSession::isApproved).count();
+    }
+
```
nsUserSessions에 새로운 수강자를 추가하고, 특정 id 또는 이름으로 조회하는 메서드는 필요 없을까?

javajigi commented on src/main/java/nextstep/courses/domain/session/Session.java:
```diff
@@ -0,0 +1,155 @@
+        }
+    }
+
+    public NsUserSession enroll(Payment payment) throws CannotEnrollException {
+        sessionStatus.canEnroll();
+        sessionCondition.match(payment);
+
+        return new NsUserSession(payment.getSessionId(), payment.getNsUserId(), EnrollmentStatus.get(approvalRequired));
+    }
```
필드의 수가 많아지면서 이 객체의 복잡도가 높아질 가능성이 점점 더 높아진다.
이 객체의 핵심 로직이라 할 수 있는 수강신청 로직을 다음과 같이 별도의 객체로 분리해 보는 것은 어떨까?

```java
Enrollment enrollment = session.enrollment(nsUserSessions);
NsUserSession nsUserSession = enrollment.enroll(payment);
```

Enrollment는 수강 신청 로직 구현을 위해 필요한 Session의 일부 값을 가지고, 수강 신청 로직 구현을 전담한다.
이렇게 구현할 경우 유지보수와 테스트 코드 구현 입장에서 고민해 보면 어떨까?

javajigi commented on src/main/java/nextstep/courses/service/SessionService.java:
```diff
@@ -18,18 +25,30 @@ public class SessionService {
+        // 수강 신청을 승인 혹은 취소한 강사 정보 및 신청 정보 값이 들어온다.
+        Session session = sessionDAO
+                .findById(nsUserSession.sessionId())
+                .with(new NsUserSessions(sessionDAO.findNsUserSessionsBySessionId(nsUserSession.sessionId())));
```
매번 전체 수강생을 조회하는 것은 성능상 이슈가 될 수 있다.
현재까지 수강한 수강생 수를 Session 테이블에 역정규화해 가지고 있는 것은 어떨까?

javajigi reviewed (changes requested):
마지막 미션 4단계까지 구현 잘 했네요. 💯 
마무리하려다 리뷰어가 바뀐 입장에서 저의 피드백 몇 개 남겨봤어요.
미션 진행할 때 너무 디자인 패턴에 집중하지는 않고 도전해 봐도 좋을 것 같아요

Perhona committed:
docs(step4): 피드백 사항 추가

Perhona committed:
refactor(step4): NsUserSession 관련 클래스명 변경, EnrollmentStatus 역할 추가

- NsUserSession 관련 테이블명 변경

Perhona committed:
refactor(step4): session이 수강인원을 직접 관리하도록 변경

- 테이블 컬럼 추가
- update userNumber 메서드 추가

Perhona committed:
refactor(step4): 수강신청 승인/취소를 Student가 담당하도록 변경

- 수강신청 승인/취소 테스트 위치 변경
- Session의 수강인원 직접 관리에 따른 관련 Students 메서드 삭제

Perhona committed:
refactor(step4): 수강 신청을 전담하는 클래스 Enrollment 추가

- 기존 Session에 있던 테스트 메서드 분리
- TeacherDoesNotMatchException 추가

Perhona committed:
refactor(step4): sessionId, nsUserId로 수강신청 리스트를 조회하는 메서드 추가, 변수에 final 선언

Perhona commented:
포비님 안녕하세요! 피드백 주신 사항 반영해보았습니다.

반영하면서 궁금했던 사항이 있어 질문드립니다.

1. Enrollment 클래스 분리 관련
> 클래스 분리를 하고 보니, 각 객체들은 본인의 작은 메서드만을 테스트하면 되어 코드가 간결하고 테스트가 쉬워졌습니다.
>  (일례로, 수강신청 테스트 시 강사 정보 확인, 인원 초과 여부까지 확인하던 부분을 삭제하고 각 클래스에서 해당 메서드만을 테스트하도록 변경하였습니다.)
> 수강신청 관련 요구사항이 변경될 경우엔 Enrollment의 코드를 변경해주면 되니 유지보수 측면에서도 용이할 것이라 생각됩니다.
> 제가 느낀 부분이 알려주려고 하신 내용이 맞는지 궁금합니다.

2.  Student의 updateEnrollmentStatus() 메서드의 인자가 3개인 문제
> Student가 자신의 상태를 관리하려다 보니, 해당 메서드에 인자를 3개 받게 되었습니다. 
> 서비스단에서 session의 강사와 일치하는지 분리하고자 했지만 결국 해당 역할 또한 상태를 변경하는 기능에 속한다고 생각하여 그대로 두었습니다.
> 객체지향 원칙에서 메서드의 인자는 되도록 2개가 넘지 않도록 하라고 배웠는데, 더 좋은 방법이 있을지 궁금합니다.

3. Students 에 새로운 수강자를 추가하고, id 등으로 찾는 메서드의 용도
> 위에 피드백 주신 내용이었는데, 어떤 경우에 사용이 될 지 이해가 잘 되지 않아 students에 메서드만 추가하였습니다.
> 제가 생각하는 흐름으로는, 수강 신청이 승인이 되면 student 테이블에 데이터가 추가되고 필요 시 강의의 수강신청 정보, 혹은 유저의 수강신청 정보를 불러오게 될 것 같은데 다른 뜻이셨는지 궁금합니다.

javajigi commented on src/main/java/nextstep/courses/domain/Students.java:
```diff
@@ -0,0 +1,32 @@
+import java.util.List;
+import java.util.Optional;
+
+public class Students {
```
👍

javajigi reviewed (approved):
피드백 반영 잘 했네요. 💯 
마지막 미션 마무리할께요.
지금까지 넘 수고했어요.

질문과 관련한 답변입니다.
1. 객체를 분리함으로써 얻는 잇점을 정확하게 파악했네요.
앞으로도 도메인 객체의 로직의 복잡도가 높아지면 새로운 객체로 분리하는 것도 고려해 보면 좋겠어요.

2. 인자수가 많아진다면 인자를 묶어 새로운 객체로 분리하는 것도 좋은 방법이지 않을까요?
1번을 통해 느낀 부분과 비슷한 시도를 해볼 수도 있다 생각해요.

3. 일급 콜렉션의 경우 콜렉션의 값을 외부에서 조작하고, 조회하는 경우가 있는데요.
지금 구현 단계에서 필요한 부분이 없다면 굳이 추가할 필요는 없을 것 같아요.
